src/main/java/org/apache/sling/discovery/base/commons/BaseDiscoveryService.java

```diff
@@ -93,7 +99,37 @@ public TopologyView getTopology() {
                 .listInstances(localClusterView);
         topology.addInstances(attachedInstances);
 
+        // Check if topology changes should be delayed
+        if (topologyReadinessHandler != null && topologyReadinessHandler.shouldDelayTopologyChange(null)) {
+            logger.debug("getTopology: topology changes are delayed, returning old view");
+            return oldView;
+        }
+
         return topology;
     }
 
+    protected void handleTopologyEvent(TopologyEvent event) {
+        if (event == null) {
+            return;
+        }
+
+        if (topologyReadinessHandler != null) {
+            if (topologyReadinessHandler.shouldDelayTopologyChange(event)) {
+                logger.debug("handleTopologyEvent: delaying topology event: {}", event);
+                return;
+            }
+
+            if (event.getType() == Type.TOPOLOGY_CHANGING) {
+                topologyReadinessHandler.startTopologyChange();
+            } else if (event.getType() == Type.TOPOLOGY_CHANGED) {
+                topologyReadinessHandler.endTopologyChange();
+            }
+        }
+
+        // Update old view when topology changes
+        if (event.getType() == Type.TOPOLOGY_CHANGED && event.getNewView() != null) {
+            setOldView((DefaultTopologyView) event.getNewView());
+        }
```

Comment:
What is the intention of this part? It looks problematic, it is not the expectation that a TopologyEventListener modifies the DiscoveryService's state - and it is actually already called by subclasses of BaseDiscoveryService. I think this is not needed (unless I'm missing something)...